ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/strategy/PutStreamerStrategy.java

```diff
@@ -0,0 +1,123 @@
+
+    private int retryAfterLeft;
+
+
+
+    public PutStreamerStrategy(final Ds3Client client, final MasterObjectList masterObjectList, final int retryAfter, final int retryDelay, final ChunkEventHandler chunkEventHandler) {
+        super(client, masterObjectList, retryAfter, retryDelay, chunkEventHandler);
+        this.filteredChunkIterator = filterChunks(masterObjectList.getObjects()).iterator();
+        this.uuidJobNodeImmutableMap = buildNodeMap(masterObjectList.getNodes());
+    }
+
+    @Override
+    public Iterable<JobPart> getWork() throws IOException, InterruptedException {
+        final Objects nextChunk = allocateChunk(filteredChunkIterator.next());
+
+        LOG.debug("Allocating chunk: {}", nextChunk.getChunkId().toString());
+        return FluentIterable.from(nextChunk.getObjects()).transform(new Function<BulkObject, JobPart>() {
+            @Nullable
+            @Override
+            public JobPart apply(@Nullable final BulkObject input) {
+
+                return new JobPart(StrategyUtils.getClient(uuidJobNodeImmutableMap,nextChunk.getNodeId(), getClient()), input);
+            }
+        });
+    }
+
+    @Override
+    public void blobCompleted(final BulkObject bulkObject) {
+
+    }
+
+    private Objects allocateChunk(final Objects filtered) throws IOException {
+        Objects chunk = null;
+        while (chunk == null) {
+            chunk = tryAllocateChunk(filtered);
+        }
+        return chunk;
+    }
+
+    private Objects tryAllocateChunk(final Objects filtered) throws IOException {
+        final AllocateJobChunkSpectraS3Response response =
+                getClient().allocateJobChunkSpectraS3(new AllocateJobChunkSpectraS3Request(filtered.getChunkId().toString()));
+
+        LOG.info("AllocatedJobChunkResponse status: {}", response.getStatus().toString());
+        switch (response.getStatus()) {
+        case ALLOCATED:
+            retryAfterLeft = getRetryAfter(); // Reset the number of retries to the initial value
+            return response.getObjectsResult();
+        case RETRYLATER:
+            try {
+                if (getRetryAfter() != -1 && retryAfterLeft == 0) {
+                    throw new Ds3NoMoreRetriesException(getRetryAfter());
+                }
+                retryAfterLeft--;
+
+                final int retryAfter = computeRetryAfter(response.getRetryAfterSeconds());
+                getChunkEventHandler().emitWaitingForChunksEvents(retryAfter);
+
+                LOG.debug("Will retry allocate chunk call after {} seconds", retryAfter);
+                Thread.sleep(retryAfter * 1000);
+                return null;
+            } catch (final InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        default:
+            assert false : "This line of code should be impossible to hit."; return null;
+        }
+    }
+
+    private int computeRetryAfter(final int retryAfterSeconds) {
```

Comment:
This is the same function as `computeDelay` in `GetStreamerStrategy`.  Reused code should be moved to base class or utility.